components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,394 @@
+                            file_dict.files.insert(file_name, info);
+                        }
+                        LogRecord::REMOVE => {
+                            let original = file_dict.files.remove(&file_name);
+                            if original.is_none() {
+                                return Err(box_err!(
+                                    "Try to recovery from log file but remove a null entry, file name: {}",
+                                    file_name
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&mut self, name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&mut self, name: &str) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::REMOVE)?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Replace existed file entry with new file information.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn replace(&mut self, src_name: &str, dst_name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes1 = Self::convert_record_to_bytes(src_name, LogRecord::REMOVE)?;
+            let bytes2 = Self::convert_record_to_bytes(dst_name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes1)?;
+            file.write_all(&bytes2)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// This function needs to be called after each append operation to check
+    /// if compact is needed.
+    fn check_compact(&mut self) -> Result<()> {
+        self.append_num += 1;
+        if self.append_num > Self::REWRITE_THRESHOLD {
```

Comment:
What if almost all operations are append? It's unnecessary to compact it in fact.

My suggestion is use `append_op_count` and `remove_op_count` to determine whether it's necessary to compact it or not. Maybe it's better to use `bytes` instead of `count`.

Reply:
how about just use length of internal `file_dict` in `LogFile`

Reply:
You need to consider the case that the disk file is very large, but the file dictionary is very small.

Reply:
Now using `self.removed` to indicate if needs to GC